animalcontrol.lua

```diff
@@ -0,0 +1,211 @@
+                           valid values: slaughter, gelding
+
+other options:
+    help                 - Displays this information
+
+]====]
+
+local Units = df.global.world.units.active
+
+if args.race and not tonumber(args.race) then
+    args.race=string.upper(args.race)
+    local raceID
+    for i,c in ipairs(df.global.world.raws.creatures.all) do
+      if c.creature_id == args.race then
+        raceID = i
+        break
+      end
+    end
+    if not raceID then
+      qerror('Invalid race: ' .. args.race)
+    end
+    args.race = raceID
+end
+
+function safe_pairs(item, keys_only)
+    if keys_only then
+        local mt = debug.getmetatable(item)
+        if mt and mt._index_table then
+            local idx = 0
+            return function()
+                idx = idx + 1
+                if mt._index_table[idx] then
+                    return mt._index_table[idx]
+                end
+            end
+        end
+    end
+    local ret = table.pack(pcall(function() return pairs(item) end))
+    local ok = ret[1]
+    table.remove(ret, 1)
+    if ok then
+        return table.unpack(ret)
+    else
+        return function() end
+    end
+end
+bfilters=(args.id or args.race or args.markedfor or args.notmarkedfor or args.gelded or args.notgelded or args.male or args.female)
+bcommands=(args.showstats or args.markfor or args.unmarkfor)
+bvalid=(args.all and not bfilters) or (not args.all and (bfilters or bcommands))
+
+if args.help or not bvalid then
+    print(help)
+else
+    count=0
+    if args.showstats then
+        print(string.format("%-20s %-6s %-9s %-9s %-5s %-22s %-8s %-25s %-7s %-7s %-7s %-7s %-7s %-7s"
+            , "animal type", "id", "unit id", "race id", "sex", "marked for slaughter", "gelded", "marked for gelding"
+            ,"str","agi","tgh","endur","recup","disres"))
+    else
+        print(string.format("%-20s %-6s %-9s %-9s %-5s %-22s %-8s %-25s"
+            , "animal type", "id", "unit id", "race id", "sex", "marked for slaughter", "gelded", "marked for gelding"))
+    end
+    for k,v in safe_pairs(Units) do
+        if v.civ_id == df.global.ui.civ_id and v.flags1.tame then
+            if not (args.male or args.female) or args.male and v.sex == 1 or args.female and v.sex == 0 then
+                if not args.race or tonumber(args.race) == v.race then
+                    if not args.markedfor or (args.markedfor == "slaughter" and v.flags2.slaughter) or (args.markedfor == "gelding" and v.flags3.marked_for_gelding) then
+                        if not args.notmarkedfor or (args.notmarkedfor == "slaughter" and not v.flags2.slaughter) or (args.notmarkedfor == "gelding" and not v.flags3.marked_for_gelding) then
+                            if not args.gelded or v.flags3.gelded then
+                                if not args.notgelded or not v.flags3.gelded then
+                                    if not args.id or tonumber(args.id) == v.id then
```

Comment:
You might want to consider moving these checks into a separate function - 8 nested conditions is rather a lot. Could be something like this (noting that my checks are inverted from yours):
```lua
function canIncludeUnit(args, v)
    if args.race and tonumber(args.race) ~= v.race then 
        return false
    elseif args.markedfor == "slaughter" and not v.flags2.slaughter then
        return false
    ....
    elseif args.gelded and not v.flags3.gelded then
        return false
    ...
    else
        return true
    end
end
```